pkg/client/cache/fifo.go

```diff
@@ -22,12 +22,15 @@ import (
 	"k8s.io/kubernetes/pkg/util/sets"
 )
 
+type PopProcessFunc func(interface{}) error
+
 // Queue is exactly like a Store, but has a Pop() method too.
 type Queue interface {
 	Store
 
-	// Pop blocks until it has something to return.
-	Pop() interface{}
+	// Pop blocks until it has something to process.
+	// It returns the object that was process and the result of processing.
+	Pop(PopProcessFunc) (interface{}, error)
```

Comment:
Unfortunately that's not going to happen. Basically the problem is that we need to sync between DeltaFIFO and "knownKeys" that is passed to it. If we don't change the interface like that, the locking will be spreaded all over the code everywhere.